File: python-DSA/mergesort.c
#include<stdio.h>
#include<stdlib.h>

void merge(int *left, int *right, int *list, int nL, int nR) {
    int i = 0, j = 0, k = 0;
    while (i < nL && j < nR) {
        if (left[i] <= right[j]) {
            list[k++] = left[i++];
        } else {
            list[k++] = right[j++];
        }
    }
    while (i < nL) {
        list[k++] = left[i++];
    }
    while (j < nR) {
        list[k++] = right[j++];
    }
}

void mergeSort(int *list, int n) {
    if (n <= 1) {
        return;
    }
    int mid = n / 2;
    int *left = malloc(sizeof(int) * mid);
    int *right = malloc(sizeof(int) * (n - mid));
    for (int i = 0; i < mid; i++) {
        left[i] = list[i];
    }
    for (int i = 0; i < n - mid; i++) {
        right[i] = list[i + mid];
    }
    mergeSort(left, mid);
    mergeSort(right, n - mid);
    merge(left, right, list, mid, n - mid);
    free(left);
    free(right);
}

int main() {
    int list[] = {38, 27, 43, 3, 9, 82, 10};
    int n = sizeof(list) / sizeof(list[0]);
    mergeSort(list, n);
    for (int i = 0; i < n; i++) {
        printf("%d ", list[i]);
    }
    printf("\n");
    return 0;
}

File: python-DSA/mergesort.py
def mergesort(list):
    """
    Sorts a list in ascending order

    Divide: find the midpoint of the list and divide into sublists

    Conquer: Recursively sort the sublists created

    Combine:Merge the sorted sublists
    """

    if len(list)<=1:
        return list

    left,right = split(list)
    left = mergesort(left)
    right = mergesort(right)

    return merge(left,right)

def split(list):
    """
    divides the unsorted list into 2 halves and returns the left and right sublists
    """
    midpoint = len(list)//2
    lefthalf = list[:midpoint]
    righthalf = list[midpoint:]
    return(lefthalf, righthalf)

def merge(left,right):
    """
    function that merges 2 lists, sorting them in the process
    """
    i=0
    j=0
    l=list()


    while i<len(left) and j<len(right):
        if left[i]<right[j]:
            l.append(left[i])
            i+=1
        else:
            l.append(right[j])
            j+=1

    while i<len(left):
        l.append(left[i])
        i+=1

    while j<len(right):
        l.append(right[j])
        j+=1

    return l

def verify_sorted(list):
    n = len(list)
    if n==0 or n==1:
        return True
    else:
        return list[0]<=list[1] and verify_sorted(list[1:])

l = [3,6,8,2,4,1,46464738,69,342,45363,4545,735,4]
l1 = mergesort(l)
print(mergesort(l1))
print(verify_sorted(l1))




def check(num1,num2):
    return num1<num2





